backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,80 @@
+package com.javahelp.backend.domain.user.register;
+
+import com.javahelp.backend.data.IUserStore;
+import com.javahelp.model.user.User;
+import com.javahelp.model.user.UserInfo;
+import com.javahelp.model.user.UserPassword;
+
+/**
+ * An interactor to make new {@link User} and {@link UserPassword} for all types of users
+ */
+public abstract class UserRegisterInteractor <T extends IUserRegisterInputBoundary,
+        U extends UserInfo>{
+
+    IUserStore userStore;
+
+    /**
+     * Creates a new {@link UserRegisterInteractor} instance
+     *
+     * @param userStore  {@link IUserStore} to use
+     */
+    public UserRegisterInteractor(IUserStore userStore) {
+        this.userStore = userStore;
+    }
+
+    /**
+     * Checking for violations for registering any {@link User} account.
+     * @param boundary an instance of {@link IUserRegisterInputBoundary} to retrieve input info
+     *
+     * @return a {@link RegisterUserResponse} model that contains message of failure at registration
+     * Null if all tests passed.
+     */
+    protected RegisterUserResponse areEmailAndUsernameTaken(T boundary){
+        if ((userStore.readByEmail(boundary.getEmailAddress()) != null)
+                && (userStore.readByUsername(boundary.getUsername()) != null)){
+            // If both username and email address exist, then return both errors
+            return new RegisterUserResponse("Both the given email address and username " +
+                    "are already occupied");
+        }
+        if (IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null){
+            return new RegisterUserResponse("The email address is already occupied");
+        }
+        else if (IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null){
```

Comment:
You shouldn't be suing `IUserStore.getDefaultImplementation` here. You should be using the class' `userStore` field. Also, instead of running these queries 4 times, you can just run them twice and store the results from the first two. There is no reason to fetch the same information twice for each of these, it will slow down the process by requiring twice as many database calls as necessary.